beacon-chain/db/kv/custody.go

```diff
@@ -38,23 +41,63 @@ func (s *Store) UpdateCustodyInfo(ctx context.Context, earliestAvailableSlot pri
 			storedEarliestAvailableSlot = primitives.Slot(bytesutil.BytesToUint64BigEndian(storedEarliestAvailableSlotBytes))
 		}
 
-		// Exit early if the new custody group count is lower than or equal to the stored one.
-		if custodyGroupCount <= storedGroupCount {
-			return nil
-		}
+		originalStoredGroupCount := storedGroupCount
 
-		storedGroupCount, storedEarliestAvailableSlot = custodyGroupCount, earliestAvailableSlot
-
-		// Store the earliest available slot.
-		bytes := bytesutil.Uint64ToBytesBigEndian(uint64(earliestAvailableSlot))
-		if err := bucket.Put(earliestAvailableSlotKey, bytes); err != nil {
-			return errors.Wrap(err, "put earliest available slot")
+		// Update custody group count only if it increased.
+		if custodyGroupCount > storedGroupCount {
+			storedGroupCount = custodyGroupCount
+			bytes := bytesutil.Uint64ToBytesBigEndian(custodyGroupCount)
+			if err := bucket.Put(groupCountKey, bytes); err != nil {
+				return errors.Wrap(err, "put custody group count")
+			}
 		}
 
-		// Store the custody group count.
-		bytes = bytesutil.Uint64ToBytesBigEndian(custodyGroupCount)
-		if err := bucket.Put(groupCountKey, bytes); err != nil {
-			return errors.Wrap(err, "put custody group count")
+		// Update earliest available slot if it advanced.
+		// This is for pruning to work correctly as blocks are pruned,
+		// the earliest available slot moves forward independently of custody group changes.
+		//
+		// IMPORTANT: If we're increasing earliestAvailableSlot without also increasing
+		// custodyGroupCount, we must ensure the new slot doesn't exceed the minimum
+		// required slot (based on MIN_EPOCHS_FOR_BLOCK_REQUESTS from current time).
+		// This prevents nodes from arbitrarily refusing to serve mandatory historical data.
+		if earliestAvailableSlot > storedEarliestAvailableSlot {
+			// If custody group count is NOT increasing, validate the increase is allowed
```

Comment:
It's better to invert the `if` condition (and return in the `if`) to reduce the indentation.

Reply:
What happens in case of backfill? Is the stored `earliestAvailableSlot` value decreased in the DB?